socket-util: introduce CMSG_FIND_AND_COPY_DATA()

yuwata opened the pull request:
Fixes #27241.

poettering commented:
Hmm, what? struct timeval is not properly aligned when included in cmsg on riscv32? That sounds completely broken and a toolchain/libc bug.

yuwata commented:
Looking at the `CMSG_XYZ()` macros in glibc, the alignment of data is only for `size_t` (I guess that's equivalent to `uintptr_t`). So, I am not sure this is a bug in glibc, but, at least `CMSG_XYZ` macros do not guarantee that the timespec is aligned with its `__alignof__()`.

poettering commented:
what a bad interface of the kernel there! meh!

i guess this triggered only on riscv32 so far because there you have 32bit pointers but 64bit time_t, which no other arch appears to have

poettering commented:
we should probably be copying out struct ucred too, just to be safe...

there's another place we deal with cmsg/SO_TIMESTAMP: server_process_datagram() in journal/journald-server.c.

poettering commented:
> we should probably be copying out struct ucred too, just to be safe...

hmm, let's not do that, instead check the alignment at compile time. will prep patch for that part.

poettering commented:
I prepped #27254 now which extends the alignment checks in CMSG_TYPED_DATA() to be done at build-time already. THis should catch the original issue during building already on riscv32 instead of runtime.

poettering commented:
PR looks conceptually good to me, just needs to cover this other struct timeval/struct timespec cases too, as mentioned.

yuwata commented:
Updated. PTAL.
Now this is on top of #27254. Please merge #27254 first.

bluca commented on src/basic/socket-util.c:
```diff
@@ -1171,6 +1171,18 @@ struct cmsghdr* cmsg_find(struct msghdr *mh, int level, int type, socklen_t leng
+        if (!cmsg)
+                return NULL;
+
+        return memcpy_safe(buf, CMSG_DATA(cmsg), buf_len == SIZE_MAX ? cmsg->cmsg_len : buf_len);
```
please add a comment here explaining why this is necessary - quoting the manpage is enough

bluca replied:
still unaddressed?

yuwata replied:
Do you think so? What kind of message do you want ?

bluca replied:
I meant to add it as a code comment, as it's obscure enough so it's bound to look puzzling in the future when one has to look here, not just as a commit message. A short version of the commit message would be more than enough.

yuwata replied:
OK. I will do.

poettering commented:
looks good but the last patch doesn't work

yuwata commented on src/libsystemd-network/sd-dhcp-server.c:
```diff
@@ -407,7 +407,7 @@ static int dhcp_server_send_udp(sd_dhcp_server *server, be32_t destination,
                    when we gain support for arbitrary number of server addresses
                  */
-                pktinfo = (struct in_pktinfo*) CMSG_DATA(cmsg);
+                pktinfo = CMSG_TYPED_DATA(cmsg, struct in_pktinfo);
```
Hm? I cannot follow. 

`CMSG_TYPED_DATA()` is mostly a typesafe version of `CMSG_DATA()` with alignment check. And no extra values are read from the cmsg header other than `CMSG_DATA()` reads. If it does not work, then I think `CMSG_DATA()` here cannot work.

> That's because it validates the size values in the cmsg header, but those are not initialized here.

Here "the size values" mean `cmsg->cmsg_len` ? It _is_ already initialized before `CMSG_TYPED_DATA()` (or `CMSG_DATA()`) is called.

yuwata commented:
Rebased, addressed @medhefgo's comment in the previous PR, extended commit message.

I hope all comments except for the @poettering's comment about the last commit.

PTAL.

dtardon commented on src/basic/socket-util.h:
```diff
@@ -189,11 +189,16 @@ int flush_accept(int fd);
         CMSG_TYPED_DATA(cmsg_find(mh, level, type, CMSG_LEN(sizeof(ctype))), ctype)
 
+/* Type-safe version of cmsg_find_and_copy_data() */
+#define CMSG_FIND_AND_COPY_DATA(mh, level, cmsg_type, type)             \
```
Nitpick: It would be nice if `CMSG_FIND_DATA` and `CMSG_FIND_AND_COPY_DATA` used same names for the type parameters.

kraj commented:
I tested this along with 27254 on 253 release and can confirm that issue is fixed on qemu riscv32

```
root@qemuriscv32:~# /usr/lib/systemd/systemd-timesyncd
Contacted time server 216.239.35.0:123 (time1.google.com).

```

yuwata commented:
Also commented in source code. Upgrading the green label.

bluca commented:
This is a bit too invasive for a stable backport
